### PR TITLE
Support "causal" padding in tf.layers.convolutional.Conv1D

### LINKED ISSUE #14933: Support "causal" padding in tf.keras.layers.Conv1D by adding support to tf.layers.convolutional.Conv1D
TensorFlow version:1.4.0

In the API docs, tf.keras.layers.Conv1D supports "causal". However, if I use "causal" as the argument of 'padding', there will be an error like this: The 'padding' argument must be one of" valid "," same ". Received : causal.

I check the source code, the Conv1D uses the methods in tensorflow.python.layers.convolutional, but the default implementations of tensorflow do not support  "causal" as the padding.

(p.s., fchollet/keras uses the conv1d method in tensorflow_backend.py)

### PR DESCRIPTION
**Pull request by facaiy**
Because #15000 is closed by mistake, hence the PR is reopened to resolve #14933 issue.

Because we don't see causal padding in other use cases expect of NTC, we choose to modify code at Conv1D, instead of tf.nn.convolution.

Ref: [conv1d implementation](https://github.com/fchollet/keras/blob/d956d19fccf6de6344c282218f1b027453785fa9/keras/backend/tensorflow_backend.py#L3141-3145) in keras.

### How to test

+ [x] add test for layers.Conv1D
+ [x] add test for keras.layers.Conv1D
+ [ ] pass all tests.

### CODE REVIEW

**Comment by facaiy** [comment]
cc @martinwicke @lukaszkaiser @fchollet who I think might be interested.

File: tensorflow/python/layers/convolutional.py [hunk 1]
```diff
@@ -290,7 +295,14 @@ def __init__(self, filters,
                trainable=True,
                name=None,
                **kwargs):
-    super(Convolution1D, self).__init__(
+    if padding == "causal":
```
Comment (facaiy) [inline tensorflow/python/layers/convolutional.py]: @szpssky note the code block is before invoking `super(Conv1D, self).__init__(`.  In your case, `causal` leaks into its parent class. I think there might be two reasons:
1. You might paste the code into wrong place by mistake, or
2. The multiple inheritance order of python varies on different environment, or I misunderstand the rule.

**Review by fchollet — CHANGES_REQUESTED** [review]
- Use `'` as string delimiter for consistency, everywhere
- `layer.padding` should return "causal". Currently you're encoding causal padding via `layer.padding == valid + layer._causal_padding == True`, which is quite confusing.
- Instead of having a method `_helper_for_preprocess` that returns two functions, have a method `_left_causal_padding`, and incorporate the code for output shape computation in `_compute_output_shape`.

**Comment by facaiy** [comment]
Thanks for review, @fchollet . I agree that `self.padding` should be `causal`, and there are two ways I can think:
1. Modify code of `_Conv` and `Conv1D`: only support causal for NTC format. Easy, but a little tricky.
2. As suggested @lukaszkaiser in #15000,  Support LEFT padding in `tf.nn.convolution`: More general.
> In general, it'd be very very good to have causal="LEFT" padding in TF. This is the one reason why both T2T and Sonet re-implement conv layers...

I think the second way sounds unified. What do you think about the idea?

**Comment by fchollet** [comment]
We can add `padding='causal'` in `tf.nn.convolution` indeed. Having support only in the layers in a bit restricted (although it is still pretty good, since most people just use the layers).

**Comment by facaiy** [comment]
@fchollet I'd like to implement the specific case at first, 

Reply:
+ Clean double quote: done.
+ `layer.padding` should return "causal": done.
     create a private method `_get_padding`, which can be removed if `nn_ops.convolution` support causal padding in the future.
+ incorporate the code for output shape computation: not yet.
    Because both `build` and `_compute_out_shape` need correct `input_shape`, I cannot find an appropriate way to incorporate. So I leave the `inputs_fn` and `shape_fn`. The solution might look awkward, however it doesn't modify `_Conv` codes and underline the preprocessing. I think it's clear for understand and a little change for `_Conv`.

Others:
+ Leave comments: remind us to remove those codes if `nn_ops.convolution` support causal padding in the future.
+ Handle unknown dimension case, thanks @szpssky .

If everyone has a good idea, don't hesitate to tell me. Thanks very much.

**Comment by drpngx** [comment]
@fchollet WDYT?

**Comment by martinwicke** [comment]
@fchollet WDYT?

**Comment by facaiy** [comment]
Close it since it has been a long time without any feedback.
